query/src/sql/planner/binder/sort.rs

```diff
@@ -46,91 +47,114 @@ pub struct OrderItem<'a> {
-                // from clause.
-                let column = from_context.resolve_column(database_name.as_ref().map(|v| v.name.as_str()), table_name.as_ref().map(|v| v.name.as_str()), ident).and_then(|v| {
-                    if distinct {
-                        Err(ErrorCode::SemanticError(order.expr.span().display_error("for SELECT DISTINCT, ORDER BY expressions must appear in select list".to_string())))
-                    } else {
-                        Ok(v)
+                    // If there isn't a matched alias in select list, we will fallback to
+                    // from clause.
+                    let column = from_context.resolve_column(database_name.as_ref().map(|v| v.name.as_str()), table_name.as_ref().map(|v| v.name.as_str()), ident).and_then(|v| {
+                        if distinct {
+                            Err(ErrorCode::SemanticError(order.expr.span().display_error("for SELECT DISTINCT, ORDER BY expressions must appear in select list".to_string())))
+                        } else {
+                            Ok(v)
+                        }
+                    })?;
+                    order_items.push(OrderItem {
+                        expr: order.clone(),
+                        name: column.column_name.clone(),
+                        index: column.index,
+                        need_eval_scalar: false,
+                    });
+                }
+                Expr::Literal {
+                    lit: Literal::Integer(index),
+                    ..
+                } => {
+                    let index = *index as usize - 1;
+                    if index >= projections.len() {
+                        return Err(ErrorCode::SemanticError(order.expr.span().display_error(
+                            format!("ORDER BY position {} is not in select list", index + 1),
+                        )));
                     }
-                })?;
-                order_items.push(OrderItem {
-                    expr: order.clone(),
-                    name: column.column_name.clone(),
-                    index: column.index,
-                    need_project: false,
-                });
-            } else if let Expr::Literal {
-                lit: Literal::Integer(index),
-                ..
-            } = &order.expr
-            {
-                let index = *index as usize - 1;
-                if index >= projections.len() {
-                    return Err(ErrorCode::SemanticError(order.expr.span().display_error(
-                        format!("ORDER BY position {} is not in select list", index + 1),
-                    )));
+                    order_items.push(OrderItem {
+                        expr: order.clone(),
+                        name: projections[index].column_name.clone(),
+                        index: projections[index].index,
+                        need_eval_scalar: false,
+                    });
+                }
+                Expr::BinaryOp { .. } => {
```

Comment:
`order by f(number)` is not supported.

Reply:
What does `f(number)` mean?

Reply:
Other expressions need to be covered. Such as `Expr::Function` ...